File: GroupAnagram.py
class Solution(object):
    
    def groupAnagrams(self, strs):
        """
        :type strs: List[str]
        :rtype: List[List[str]]
        """

        outMap = {}
        for str in strs):
            key = ''.join(sorted(strs[i]))
            outMap.setdefault(key, []).append(strs[i])

        output = []
        for val in outMap.values():
            output.append(val)
        return output


s = Solution()
print s.groupAnagrams(["eat", "tea", "tan", "ate", "nat", "bat"])

File: GroupAnagram.java
/*
  https://leetcode.com/problems/group-anagrams/
 */
import java.util.*;

class GroupAnagram {
  public static void main(String[] args) {
    String[] words = {"eat", "tea", "tan", "ate", "nat", "bat"};
    System.out.println(new GroupAnagram().groupAnagram(words));
  }

  public List<List<String>> groupAnagram(String[] words) {
    HashMap<String, ArrayList<String>> charMapToWord = new HashMap<String, ArrayList<String>>();

    for (String word: words) {
      short[] charFreq = new short[26];
      for(int i = 0; i < word.length(); i++) {
          charFreq[word.charAt(i) - 'a']++;
      }

      charMapToWord.computeIfAbsent(Arrays.toString(charFreq), v -> new ArrayList<>())
          .add(word);
    }

    return new ArrayList<List<String>>(charMapToWord.values());
  }
}
